In [1]:
import requests
import seaborn as sns
import plotly.express as px
from matplotlib import pyplot as plt
import numpy as np
import pandas as pd
import time

In [2]:
file_prefix = '../data/schools_attica'

In [3]:
df = pd.read_csv(f'{file_prefix}.csv')
df

Κατηγορία                       Σχολείο  Κωδικός  \
0        Γυμνάσια  12ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501150   
1        Γυμνάσια  13ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501160   
2        Γυμνάσια  14ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501170   
3        Γυμνάσια  15ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501180   
4        Γυμνάσια  16ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501190   
...           ...                           ...      ...   
3067  Νηπιαγωγεία  ΕΙΔΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΕΛΕΥΣΙΝΑΣ  9521062   
3068  Νηπιαγωγεία   ΕΙΔΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΖΕΦΥΡΙΟΥ  9521495   
3069  Νηπιαγωγεία            ΝΗΠΙΑΓΩΓΕΙΟ ΒΙΛΙΩΝ  9050620   
3070  Νηπιαγωγεία           ΝΗΠΙΑΓΩΓΕΙΟ ΕΡΥΘΡΩΝ  9050657   
3071  Νηπιαγωγεία             ΝΗΠΙΑΓΩΓΕΙΟ ΦΥΛΗΣ  9050297   

                                      Διεύθυνση  \
0            ΚΕΙΡΙΑΔΩΝ ΚΑΙ ΣΦΗΤΤΙΩΝ 1, ΤΚ 11854   
1         ΑΛΦΕΙΩΝΙΑΣ 10-12 & ΒΟΥΤΣΙΝΑ, ΤΚ 11632   
2                         ΑΝ.ΖΙΝΝΗ 28, ΤΚ 11741   
3                           ΡΗΝΕΙΑΣ 2, ΤΚ 11362   
4             Πανόρμου και Λαρίσης 47, ΤΚ 11524   
...                                         ...   
3067                      ΟΔ. ΕΛΥΤΗ 2, ΤΚ 19200   
3068                        ΣΟΥΛΙΟΥ 9, ΤΚ 13461   
3069  Μεγάλου Αλεξάνδρου & Κιθαιρώνος, ΤΚ 19012   
3070                   ΑΝΕΞΑΡΤΗΣΙΑΣ 2, ΤΚ 19008   
3071                    ΑΣΠΡΟΠΥΡΓΟΥ 6, ΤΚ 13351   

              Περιφερειακή Διεύθυνση  
0           ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ  
1           ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ  
2           ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ  
3           ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ  
4           ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ  
...                              ...  
3067  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ  
3068  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ  
3069  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ  
3070  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ  
3071  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ  

[3072 rows x 5 columns]

In [4]:
# remove the ΤΚ from the postal code
df['Διεύθυνση'] = df['Διεύθυνση'].str.replace('ΤΚ', '')

In [5]:
# replace multiple spaces with a single space
df['Διεύθυνση'] = df['Διεύθυνση'].str.replace(' +', ' ')

In [6]:
df

Κατηγορία                       Σχολείο  Κωδικός  \
0        Γυμνάσια  12ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501150   
1        Γυμνάσια  13ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501160   
2        Γυμνάσια  14ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501170   
3        Γυμνάσια  15ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501180   
4        Γυμνάσια  16ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501190   
...           ...                           ...      ...   
3067  Νηπιαγωγεία  ΕΙΔΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΕΛΕΥΣΙΝΑΣ  9521062   
3068  Νηπιαγωγεία   ΕΙΔΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΖΕΦΥΡΙΟΥ  9521495   
3069  Νηπιαγωγεία            ΝΗΠΙΑΓΩΓΕΙΟ ΒΙΛΙΩΝ  9050620   
3070  Νηπιαγωγεία           ΝΗΠΙΑΓΩΓΕΙΟ ΕΡΥΘΡΩΝ  9050657   
3071  Νηπιαγωγεία             ΝΗΠΙΑΓΩΓΕΙΟ ΦΥΛΗΣ  9050297   

                                    Διεύθυνση          Περιφερειακή Διεύθυνση  
0            ΚΕΙΡΙΑΔΩΝ ΚΑΙ ΣΦΗΤΤΙΩΝ 1,  11854        ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ  
1         ΑΛΦΕΙΩΝΙΑΣ 10-12 & ΒΟΥΤΣΙΝΑ,  11632        ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ  
2                         ΑΝ.ΖΙΝΝΗ 28,  11741        ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ  
3                           ΡΗΝΕΙΑΣ 2,  11362        ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ  
4             Πανόρμου και Λαρίσης 47,  11524        ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ  
...                                       ...                             ...  
3067                      ΟΔ. ΕΛΥΤΗ 2,  19200  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ  
3068                        ΣΟΥΛΙΟΥ 9,  13461  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ  
3069  Μεγάλου Αλεξάνδρου & Κιθαιρώνος,  19012  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ  
3070                   ΑΝΕΞΑΡΤΗΣΙΑΣ 2,  19008  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ  
3071                    ΑΣΠΡΟΠΥΡΓΟΥ 6,  13351  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ  

[3072 rows x 5 columns]

In [7]:
overpass_url = "http://overpass-api.de/api/interpreter"

In [8]:
def get_coordinates(address):
    overpass_query = f"""
    [out:json];
    node["addr:street"="{address}"];
    out center;
    """
    response = requests.get(overpass_url, params={'data': overpass_query})
    data = response.json()

    if len(data['elements']) == 0:
        return None, None

    return data['elements'][0]['lat'], data['elements'][0]['lon']

In [9]:
for (idx, address) in df['Διεύθυνση'].tail(20).items():
    lat, lon = get_coordinates(address)
    df.loc[idx, 'lat'] = lat
    df.loc[idx, 'lon'] = lon

In [28]:
with open('../google_maps_api_key') as f:
    api_key = f.read()

In [29]:
url = f'https://maps.googleapis.com/maps/api/geocode/json'

In [35]:
# get the coordinates using google maps api
for (idx, address) in df['Διεύθυνση'].head(3).items():
# for (idx, address) in df['Διεύθυνση'].items():
    if pd.notnull(df.loc[idx, 'lat']) or pd.notnull(df.loc[idx, 'lon']):
        continue

    search_params = {
        'address': address,
        'key': api_key
    }

    response = requests.get(url, params=search_params)
    data = response.json()
#     print(data)
    try:
        lat = data['results'][0]['geometry']['location']['lat']
        lon = data['results'][0]['geometry']['location']['lng']
        df.loc[idx, 'lat'] = lat
        df.loc[idx, 'lon'] = lon
    except:
        print('Error in response')
        print('Response: ', data)
        print('Address: ', address)
        print('Moving on...')
        print("\n")

Error in response
Response:  {'error_message': 'You must enable Billing on the Google Cloud Project at https://console.cloud.google.com/project/_/billing/enable Learn more at https://developers.google.com/maps/gmp-get-started', 'results': [], 'status': 'REQUEST_DENIED'}
Address:  ΡΗΝΕΙΑΣ 2,  11362
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  nan
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  nan
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  nan
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  nan
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  ΤΣΕΜΠΕΡΛΙΔΗ 5,  19014
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  ΣΩΤΗΡΟΣ 11,  19023
Moving on...


Error in response
Response:  {'results': [], 'status'

Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  Ε. ΒΕΝΙΖΕΛΟΥ 4,  14671
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  ΚΑΡΑΘΕΟΔΩΡΗ ΚΑΙ Κ. ΟΥΡΑΝΗ 16,  11525
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  ΚΟΡΥΤΣΑΣ 29,  15521
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  Τ.Θ. 51051,  14510
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  Τ.Θ. 51051,  14510
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  nan
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  ΟΜΟΡΦΟΚΛΗΣΙΑΣ 21,  15122
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  nan
Moving on...


Error in response
Response:  {'results': [], 'status': 'ZERO_RESULTS'}
Address:  

<IPython.core.display.Javascript object>

In [40]:
df

Κατηγορία                       Σχολείο  Κωδικός  \
0        Γυμνάσια  12ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501150   
1        Γυμνάσια  13ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501160   
2        Γυμνάσια  14ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501170   
3        Γυμνάσια  15ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501180   
4        Γυμνάσια  16ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   501190   
...           ...                           ...      ...   
3067  Νηπιαγωγεία  ΕΙΔΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΕΛΕΥΣΙΝΑΣ  9521062   
3068  Νηπιαγωγεία   ΕΙΔΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΖΕΦΥΡΙΟΥ  9521495   
3069  Νηπιαγωγεία            ΝΗΠΙΑΓΩΓΕΙΟ ΒΙΛΙΩΝ  9050620   
3070  Νηπιαγωγεία           ΝΗΠΙΑΓΩΓΕΙΟ ΕΡΥΘΡΩΝ  9050657   
3071  Νηπιαγωγεία             ΝΗΠΙΑΓΩΓΕΙΟ ΦΥΛΗΣ  9050297   

                                    Διεύθυνση          Περιφερειακή Διεύθυνση  \
0            ΚΕΙΡΙΑΔΩΝ ΚΑΙ ΣΦΗΤΤΙΩΝ 1,  11854        ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ   
1         ΑΛΦΕΙΩΝΙΑΣ 10-12 & ΒΟΥΤΣΙΝΑ,  11632        ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ   
2                         ΑΝ.ΖΙΝΝΗ 28,  11741        ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ   
3                           ΡΗΝΕΙΑΣ 2,  11362        ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ   
4             Πανόρμου και Λαρίσης 47,  11524        ΔΙΕΥΘΥΝΣΗ Δ.Ε. Α΄ ΑΘΗΝΑΣ   
...                                       ...                             ...   
3067                      ΟΔ. ΕΛΥΤΗ 2,  19200  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ   
3068                        ΣΟΥΛΙΟΥ 9,  13461  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ   
3069  Μεγάλου Αλεξάνδρου & Κιθαιρώνος,  19012  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ   
3070                   ΑΝΕΞΑΡΤΗΣΙΑΣ 2,  19008  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ   
3071                    ΑΣΠΡΟΠΥΡΓΟΥ 6,  13351  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ   

            lat        lon  
0     37.974195  23.711097  
1     37.957304  23.744174  
2     37.963601  23.723498  
3           NaN        NaN  
4      37.99344  23.761897  
...         ...        ...  
3067  37.907589  23.719371  
3068  37.932442  23.743177  
3069  38.153709  23.343673  
3070       None       None  
3071  38.101987  23.667161  

[3072 rows x 7 columns]

In [41]:
df.isna().sum()

Κατηγορία                   0
Σχολείο                     0
Κωδικός                     0
Διεύθυνση                  61
Περιφερειακή Διεύθυνση      0
lat                       139
lon                       139
dtype: int64

In [42]:
len(df)

3072

In [43]:
pd.set_option('display.max_rows', 100)

In [44]:
df[(df['lat'].isna()) & (~df['Διεύθυνση'].isna())]

Κατηγορία                                            Σχολείο  \
3             Γυμνάσια                       15ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΘΗΝΩΝ   
208           Γυμνάσια                  1ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΚΑΠΑΝΔΡΙΤΙΟΥ   
227           Γυμνάσια  2ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΜΑΡΚΟΠΟΥΛΟΥ ΜΕΣΟΓΑΙΑΣ ΑΤΤ...   
231           Γυμνάσια                               3ο ΓΥΜΝΑΣΙΟ ΚΟΡΩΠΙΟΥ   
332             Λύκεια                          2ο ΓΕΝΙΚΟ ΛΥΚΕΙΟ ΚΟΡΩΠΙΟΥ   
338             Λύκεια              2ο ΗΜΕΡΗΣΙΟ ΓΕΝΙΚΟ ΛΥΚΕΙΟ ΜΑΡΚΟΠΟΥΛΟΥ   
343             Λύκεια      5ο ΗΜΕΡΗΣΙΟ ΓΕΝΙΚΟ ΛΥΚΕΙΟ ΑΧΑΡΝΩΝ - ΑΓΙΑ ΑΝΝΑ   
344             Λύκεια                  6ο ΗΜΕΡΗΣΙΟ ΓΕΝΙΚΟ ΛΥΚΕΙΟ ΑΧΑΡΝΕΣ   
352             Λύκεια                ΗΜΕΡΗΣΙΟ ΓΕΝΙΚΟ ΛΥΚΕΙΟ ΑΓ. ΣΤΕΦΑΝΟΥ   
575           Γυμνάσια               3ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΓΙΩΝ ΑΝΑΡΓΥΡΩΝ   
656             Λύκεια            2ο ΗΜΕΡΗΣΙΟ ΓΕΝΙΚΟ ΛΥΚΕΙΟ ΑΓ. ΑΝΑΡΓΥΡΩΝ   
680             Λύκεια                    7ο ΗΜΕΡΗΣΙΟ ΓΕΝΙΚΟ ΛΥΚΕΙΟ ΙΛΙΟΥ   
687             Λύκεια            ΠΕΙΡΑΜΑΤΙΚΟ ΓΕΝΙΚΟ ΛΥΚΕΙΟ ΑΓ. ΑΝΑΡΓΥΡΩΝ   
700           Γυμνάσια               1ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΓΙΟΥ ΔΗΜΗΤΡΙΟΥ   
715           Γυμνάσια                     2ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΚΑΛΛΙΘΕΑΣ   
785             Λύκεια  1ο ΗΜΕΡΗΣΙΟ ΓΕΝΙΚΟ ΛΥΚΕΙΟ ΑΓΙΟΥ ΔΗΜΗΤΡΙΟΥ - ΤΑ...   
837           Γυμνάσια                       1ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΜΑΝΔΡΑΣ   
844           Γυμνάσια                   3ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΑΝΩ ΛΙΟΣΙΩΝ   
847           Γυμνάσια                       3ο ΗΜΕΡΗΣΙΟ ΓΥΜΝΑΣΙΟ ΜΕΓΑΡΩΝ   
873             Λύκεια                    ΗΜΕΡΗΣΙΟ ΓΕΝΙΚΟ ΛΥΚΕΙΟ ΖΕΦΥΡΙΟΥ   
954   Δημοτικά Σχολεία                        20ο ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ ΑΘΗΝΩΝ   
960   Δημοτικά Σχολεία                        24ο ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ ΑΘΗΝΑΣ   
1121  Ιδιωτικά Σχολεία            ΙΔΙΩΤΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΑΘΗΝΑ - ΣΧΟΛΗ ΧΙΛΛ   
1259       Νηπιαγωγεία                           19ο ΝΗΠΙΑΓΩΓΕΙΟ ΖΩΓΡΑΦΟΥ   
1278       Νηπιαγωγεία                             23ο ΝΗΠΙΑΓΩΓΕΙΟ ΑΘΗΝΩΝ   
1356       Νηπιαγωγεία                             63ο ΝΗΠΙΑΓΩΓΕΙΟ ΑΘΗΝΩΝ   
1466  Δημοτικά Σχολεία              20ο ΟΛΟΗΜΕΡΟ ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ ΑΧΑΡΝΩΝ   
1474  Δημοτικά Σχολεία              28ο ΟΛΟΗΜΕΡΟ ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ ΑΧΑΡΝΩΝ   
1527  Δημοτικά Σχολεία                4ο ΟΛΟΗΜΕΡΟ ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ ΓΕΡΑΚΑ   
1555  Δημοτικά Σχολεία                         ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ ΜΑΛΑΚΑΣΑΣ   
1668  Ιδιωτικά Σχολεία     ΠΡΟΤΥΠΗ ΕΛΛΗΝΟΓΕΡΜΑΝΙΚΗ ΑΓΩΓΗ Ο ΜΙΚΡΟΣ ΑΝΤΩΝΗΣ   
1694       Νηπιαγωγεία                            1ο ΝΗΠΙΑΓΩΓΕΙΟ ΚΕΡΑΤΕΑΣ   
1717       Νηπιαγωγεία                  1ο ΟΛΟΗΜΕΡΟ ΝΗΠΙΑΓΩΓΕΙΟ ΑΡΤΕΜΙΔΟΣ   
1768       Νηπιαγωγεία              2ο ΟΛΟΗΜΕΡΟ ΝΗΠΙΑΓΩΓΕΙΟ ΝΕΩΝ ΠΑΛΑΤΙΩΝ   
1801       Νηπιαγωγεία                             4ο ΝΗΠΙΑΓΩΓΕΙΟ ΑΧΑΡΝΩΝ   
1836       Νηπιαγωγεία            ΕΙΔΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΒΟΥΛΑ - ΕΙΔΙΚΟ ΠΙΚΠΑ   
1853       Νηπιαγωγεία                       ΟΛΟΗΜΕΡΟ ΝΗΠΙΑΓΩΓΕΙΟ ΚΟΥΒΑΡΑ   
1858  Δημοτικά Σχολεία                    10ο ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ ΑΜΑΡΟΥΣΙΟΥ   
1882  Δημοτικά Σχολεία                 1ο ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ ΝΕΑΣ ΕΡΥΘΡΑΙΑΣ   
1923  Δημοτικά Σχολεία                   3ο ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ ΝΕΟΥ ΨΥΧΙΚΟΥ   
1984  Ιδιωτικά Σχολεία  2ο ΝΗΠ/ΓΕΙΟ ΕΚΠΑΙΔΕΥΤΗΡΙΩΝ ΜΑΙΡΗΣ ΑΡΓΥΡΗ-ΛΑΙΜΟ...   
1987  Ιδιωτικά Σχολεία                     INTERNATIONAL SCHOOL OF ATHENS   
1988  Ιδιωτικά Σχολεία       INTERNATIONAL SCHOOL OF ATHENS (ΝΗΠΙΑΓΩΓΕΙΟ)   
2012  Ιδιωτικά Σχολεία    ΙΔΙΩΤΙΚΟ ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ - ΕΛΛΗΝΟΑΓΓΛΙΚΗ ΑΓΩΓΗ   
2048  Ιδιωτικά Σχολεία  ΙΔΙΩΤΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΜΑΡΟΥΣΙ - ΣΥΓΧΡΟΝΗ ΕΛΛΗΝΟ...   
2154       Νηπιαγωγεία                         1ο ΝΗΠΙΑΓΩΓΕΙΟ ΝΕΑΣ ΙΩΝΙΑΣ   
2169       Νηπιαγωγεία                          2ο ΝΗΠΙΑΓΩΓΕΙΟ ΛΥΚΟΒΡΥΣΗΣ   
2190       Νηπιαγωγεία                        3ο ΝΗΠΙΑΓΩΓΕΙΟ ΝΕΟΥ ΨΥΧΙΚΟΥ   
2203       Νηπιαγωγεία                        4ο ΝΗΠΙΑΓΩΓΕΙΟ ΝΕΟΥ ΨΥΧΙΚΟΥ   
2239       Νηπιαγωγεία                    8ο ΝΗΠΙΑΓΩΓΕΙΟ ΑΓΙΑΣ ΠΑΡΑΣΚΕΥΗΣ   
2257  Δημοτικά Σχολεία     10ο ΔΗΜΟΤΙΚΟ ΣΧΟΛΕΙΟ ΑΓΙΩΝ ΑΝΑΡΓΥΡΩΝ - ΑΝΑΚΑΣΑ   
2275  Δημ

In [45]:
# address = 'ΠΑΠΑΦΛΕΣΣΑ 6, 13461'
address = 'Papaflessa 6, 13461'

In [46]:
search_params = {
    'address': address,
    'key': api_key
}

response = requests.get(url, params=search_params)
data = response.json()

In [47]:
data

{'results': [{'address_components': [{'long_name': '6',
     'short_name': '6',
     'types': ['street_number']},
    {'long_name': 'Papaflessa',
     'short_name': 'Papaflessa',
     'types': ['route']},
    {'long_name': 'Agii Anargiri',
     'short_name': 'Agii Anargiri',
     'types': ['locality', 'political']},
    {'long_name': 'Ditikos Tomeas Athinon',
     'short_name': 'Ditikos Tomeas Athinon',
     'types': ['administrative_area_level_3', 'political']},
    {'long_name': 'Greece',
     'short_name': 'GR',
     'types': ['country', 'political']},
    {'long_name': '135 62', 'short_name': '135 62', 'types': ['postal_code']}],
   'formatted_address': 'Papaflessa 6, Agii Anargiri 135 62, Greece',
   'geometry': {'bounds': {'northeast': {'lat': 38.0346844, 'lng': 23.725197},
     'southwest': {'lat': 38.0345513, 'lng': 23.7250503}},
    'location': {'lat': 38.0346179, 'lng': 23.7251236},
    'location_type': 'ROOFTOP',
    'viewport': {'northeast': {'lat': 38.0359880802915,
      

There is a different postal code for the above address. I can only assume there is some error in the inputed data. Any address not found will have to be fixed by hand.

In [48]:
df.tail()

Κατηγορία                       Σχολείο  Κωδικός  \
3067  Νηπιαγωγεία  ΕΙΔΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΕΛΕΥΣΙΝΑΣ  9521062   
3068  Νηπιαγωγεία   ΕΙΔΙΚΟ ΝΗΠΙΑΓΩΓΕΙΟ ΖΕΦΥΡΙΟΥ  9521495   
3069  Νηπιαγωγεία            ΝΗΠΙΑΓΩΓΕΙΟ ΒΙΛΙΩΝ  9050620   
3070  Νηπιαγωγεία           ΝΗΠΙΑΓΩΓΕΙΟ ΕΡΥΘΡΩΝ  9050657   
3071  Νηπιαγωγεία             ΝΗΠΙΑΓΩΓΕΙΟ ΦΥΛΗΣ  9050297   

                                    Διεύθυνση          Περιφερειακή Διεύθυνση  \
3067                      ΟΔ. ΕΛΥΤΗ 2,  19200  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ   
3068                        ΣΟΥΛΙΟΥ 9,  13461  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ   
3069  Μεγάλου Αλεξάνδρου & Κιθαιρώνος,  19012  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ   
3070                   ΑΝΕΞΑΡΤΗΣΙΑΣ 2,  19008  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ   
3071                    ΑΣΠΡΟΠΥΡΓΟΥ 6,  13351  ΔΙΕΥΘΥΝΣΗ Π.Ε. ΔΥΤΙΚΗΣ ΑΤΤΙΚΗΣ   

            lat        lon  
3067  37.907589  23.719371  
3068  37.932442  23.743177  
3069  38.153709  23.343673  
3070       None       None  
3071  38.101987  23.667161

Address `Μεγάλου Αλεξάνδρου & Κιθαιρώνος, 19012` was found. For now, we can keep only the schools whose coordinates we know.

In [10]:
df['Κατηγορία'].value_counts()

Κατηγορία
Νηπιαγωγεία                                          926
Δημοτικά Σχολεία                                     787
Ιδιωτικά Σχολεία                                     610
Γυμνάσια                                             367
Λύκεια                                               272
Επαγγελματικά Λύκεια                                  74
Σχολικό Εργαστηριακό Κέντρο                           27
Ειδικής Επαγγελματικής Εκπαίδευσης και Κατάρτισης      9
Name: count, dtype: int64

In [24]:
df['Type'] = np.where(df['Κατηγορία'] == 'Ιδιωτικά Σχολεία', 'Private', 'Public')

In [26]:
# df.to_csv(f'{file_prefix}_coordinates.csv', index=False)

In [27]:
df = pd.read_csv(f'{file_prefix}_coordinates.csv')

In [28]:
df.isna().sum()

Κατηγορία                   0
Σχολείο                     0
Κωδικός                     0
Διεύθυνση                  61
Περιφερειακή Διεύθυνση      0
lat                       139
lon                       139
Τύπος                       0
Type                        0
dtype: int64

In [29]:
# create a large plotly express map graph
fig = px.scatter_mapbox(
    df,
    lat='lat',
    lon='lon',
    hover_name='Σχολείο',
    # color='Κατηγορία'
    )

fig.update_layout(
        mapbox_style='open-street-map',
        margin=dict(l=20, r=20, t=20, b=20),
        height=1000,
        width=1400
        )

fig.show()

In [31]:
# create a large plotly express map graph
fig = px.scatter_mapbox(
    df,
    lat='lat',
    lon='lon',
    hover_name='Σχολείο',
    color='Type'
    )

fig.update_layout(
        mapbox_style='open-street-map',
        margin=dict(l=20, r=20, t=20, b=20),
        height=1000,
        width=1400
        )

fig.show()